File: gdscript/src/main/java/gdscript/competion/utils/GdLookupElementBuilder.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.

package gdscript.competion.utils;

import com.intellij.codeInsight.completion.InsertHandler;
import com.intellij.codeInsight.completion.InsertionContext;
import com.intellij.codeInsight.lookup.AutoCompletionPolicy;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.codeInsight.lookup.LookupElementPresentation;
import com.intellij.codeInsight.lookup.LookupElementRenderer;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNamedElement;
import com.intellij.psi.SmartPointerManager;
import com.intellij.psi.SmartPsiElementPointer;
import com.intellij.psi.util.PsiUtilCore;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

/**
 * Copied directly from com.intellij.codeInsight.lookup.LookupElementBuilder
 * The only change is a new create() method with different lookup string & object -> a hack to add extra String for partial text match completions
 *
 * @author peter
 *
 * @see com.intellij.codeInsight.completion.PrioritizedLookupElement
 */
public final class GdLookupElementBuilder extends LookupElement {
    @NotNull private final String myLookupString;
    @NotNull private final Object myObject;
    @Nullable private final SmartPsiElementPointer<?> myPsiElement;
    private final boolean myCaseSensitive;
    @Nullable private final InsertHandler<LookupElement> myInsertHandler;
    @Nullable private final LookupElementRenderer<LookupElement> myRenderer;
    @Nullable private final LookupElementRenderer<LookupElement> myExpensiveRenderer;
    @Nullable private final LookupElementPresentation myHardcodedPresentation;
    @NotNull private final Set<String> myAllLookupStrings;

    private GdLookupElementBuilder(@NotNull String lookupString, @NotNull Object object, @Nullable InsertHandler<LookupElement> insertHandler,
                                 @Nullable LookupElementRenderer<LookupElement> renderer,
                                 @Nullable LookupElementRenderer<LookupElement> expensiveRenderer,
                                 @Nullable LookupElementPresentation hardcodedPresentation,
                                 @Nullable SmartPsiElementPointer<?> psiElement,
                                 @NotNull Set<String> allLookupStrings,
                                 boolean caseSensitive) {
        myLookupString = lookupString;
        myObject = object;
        myInsertHandler = insertHandler;
        myRenderer = renderer;
        myExpensiveRenderer = expensiveRenderer;
        myHardcodedPresentation = hardcodedPresentation;
        myPsiElement = psiElement;
        myAllLookupStrings = Collections.unmodifiableSet(allLookupStrings);
        myCaseSensitive = caseSensitive;
    }

    private GdLookupElementBuilder(@NotNull String lookupString, @NotNull Object object) {
        this(lookupString, object, null, null, null, null, null, Collections.singleton(lookupString), true);
    }

    public static @NotNull GdLookupElementBuilder create(@NotNull String lookupString) {
        return new GdLookupElementBuilder(lookupString, lookupString);
    }

    public static @NotNull GdLookupElementBuilder create(@NotNull String lookupString, @NotNull String lookupObject) {
        return new GdLookupElementBuilder(lookupString, lookupObject);
    }

    public static @NotNull GdLookupElementBuilder create(@NotNull Object object) {
        return new GdLookupElementBuilder(object.toString(), object);
    }

    public static @NotNull GdLookupElementBuilder createWithSmartPointer(@NotNull String lookupString, @NotNull PsiElement element) {
        PsiUtilCore.ensureValid(element);
        return new GdLookupElementBuilder(lookupString,
                SmartPointerManager.getInstance(element.getProject()).createSmartPsiElementPointer(element));
    }

    public static @NotNull GdLookupElementBuilder create(@NotNull PsiNamedElement element) {
        PsiUtilCore.ensureValid(element);
        return new GdLookupElementBuilder(StringUtil.notNullize(element.getName()), element);
    }

    public static @NotNull GdLookupElementBuilder createWithIcon(@NotNull PsiNamedElement element) {
        PsiUtilCore.ensureValid(element);
        return create(element).withIcon(element.getIcon(0));
    }

    public static @NotNull GdLookupElementBuilder create(@NotNull Object lookupObject, @NotNull String lookupString) {
        if (lookupObject instanceof PsiElement) {
            PsiUtilCore.ensureValid((PsiElement)lookupObject);
        }
        return new GdLookupElementBuilder(lookupString, lookupObject);
    }

    private @NotNull GdLookupElementBuilder cloneWithUserData(@NotNull String lookupString, @NotNull Object object,
                                                            @Nullable InsertHandler<LookupElement> insertHandler,
                                                            @Nullable LookupElementRenderer<LookupElement> renderer,
                                                            @Nullable LookupElementRenderer<LookupElement> expensiveRenderer,
                                                            @Nullable LookupElementPresentation hardcodedPresentation,
                                                            @Nullable SmartPsiElementPointer<?> psiElement,
                                                            @NotNull Set<String> allLookupStrings,
                                                            boolean caseSensitive) {
        GdLookupElementBuilder result = new GdLookupElementBuilder(lookupString, object, insertHandler, renderer, expensiveRenderer,
                hardcodedPresentation, psiElement, allLookupStrings, caseSensitive);
        copyUserDataTo(result);
        return result;
    }

    /**
     * @deprecated use {@link #withInsertHandler(InsertHandler)}
     */
    @Deprecated
    @ApiStatus.ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder setInsertHandler(@Nullable InsertHandler<LookupElement> insertHandler) {
        return withInsertHandler(insertHandler);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withInsertHandler(@Nullable InsertHandler<LookupElement> insertHandler) {
        return cloneWithUserData(myLookupString, myObject, insertHandler, myRenderer, myExpensiveRenderer, myHardcodedPresentation,
                myPsiElement, myAllLookupStrings, myCaseSensitive);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withRenderer(@Nullable LookupElementRenderer<LookupElement> renderer) {
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, renderer, myExpensiveRenderer, myHardcodedPresentation,
                myPsiElement, myAllLookupStrings, myCaseSensitive);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withExpensiveRenderer(@Nullable LookupElementRenderer<LookupElement> expensiveRenderer) {
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, myRenderer, expensiveRenderer, myHardcodedPresentation,
                myPsiElement, myAllLookupStrings, myCaseSensitive);
    }

    @Override
    @NotNull
    public Set<String> getAllLookupStrings() {
        return myAllLookupStrings;
    }

    /**
     * @deprecated use {@link #withIcon(Icon)}
     */
    @Deprecated
    @ApiStatus.ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder setIcon(@Nullable Icon icon) {
        return withIcon(icon);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withIcon(@Nullable Icon icon) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.setIcon(icon);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
                myAllLookupStrings, myCaseSensitive);
    }

    @NotNull
    private LookupElementPresentation copyPresentation() {
        final LookupElementPresentation presentation = new LookupElementPresentation();
        if (myHardcodedPresentation != null) {
            presentation.copyFrom(myHardcodedPresentation);
        } else {
            presentation.setItemText(myLookupString);
        }
        return presentation;
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withLookupString(@NotNull String another) {
        final Set<String> set = new HashSet<>(myAllLookupStrings);
        set.add(another);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, myRenderer, myExpensiveRenderer, myHardcodedPresentation,
                myPsiElement, Collections.unmodifiableSet(set), myCaseSensitive);
    }
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withLookupStrings(@NotNull Collection<String> another) {
        Set<String> set = new HashSet<>(myAllLookupStrings.size() + another.size());
        set.addAll(myAllLookupStrings);
        set.addAll(another);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, myRenderer, myExpensiveRenderer, myHardcodedPresentation,
                myPsiElement, Collections.unmodifiableSet(set), myCaseSensitive);
    }

    @Override
    public boolean isCaseSensitive() {
        return myCaseSensitive;
    }

    /**
     * @param caseSensitive if this lookup item should be completed in the same letter case as prefix
     * @return modified builder
     * @see com.intellij.codeInsight.completion.CompletionResultSet#caseInsensitive()
     */
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withCaseSensitivity(boolean caseSensitive) {
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, myRenderer, myExpensiveRenderer, myHardcodedPresentation,
                myPsiElement, myAllLookupStrings, caseSensitive);
    }

    /**
     * Allows to pass custom PSI that will be returned from {@link #getPsiElement()}.
     */
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withPsiElement(@Nullable PsiElement psi) {
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, myRenderer, myExpensiveRenderer, myHardcodedPresentation,
                psi == null ? null : SmartPointerManager.createPointer(psi),
                myAllLookupStrings, myCaseSensitive);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withItemTextForeground(@NotNull Color itemTextForeground) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.setItemTextForeground(itemTextForeground);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation,
                myPsiElement, myAllLookupStrings, myCaseSensitive);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withItemTextUnderlined(boolean underlined) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.setItemTextUnderlined(underlined);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation,
                myPsiElement, myAllLookupStrings, myCaseSensitive);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withItemTextItalic(boolean italic) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.setItemTextItalic(italic);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation,
                myPsiElement, myAllLookupStrings, myCaseSensitive);
    }

    /**
     * @deprecated use {@link #withTypeText(String)}
     */
    @Deprecated
    @ApiStatus.ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder setTypeText(@Nullable String typeText) {
        return withTypeText(typeText);
    }
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withTypeText(@Nullable String typeText) {
        return withTypeText(typeText, false);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withTypeText(@Nullable String typeText, boolean grayed) {
        return withTypeText(typeText, null, grayed);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withTypeText(@Nullable String typeText, @Nullable Icon typeIcon, boolean grayed) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.setTypeText(typeText, typeIcon);
        presentation.setTypeGrayed(grayed);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
                myAllLookupStrings, myCaseSensitive);
    }

    public @NotNull GdLookupElementBuilder withTypeIconRightAligned(boolean typeIconRightAligned) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.setTypeIconRightAligned(typeIconRightAligned);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
                myAllLookupStrings, myCaseSensitive);
    }

    /**
     * @deprecated use {@link #withPresentableText(String)}
     */
    @Deprecated
    @ApiStatus.ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder setPresentableText(@NotNull String presentableText) {
        return withPresentableText(presentableText);
    }
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withPresentableText(@NotNull String presentableText) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.setItemText(presentableText);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
                myAllLookupStrings, myCaseSensitive);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder bold() {
        return withBoldness(true);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withBoldness(boolean bold) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.setItemTextBold(bold);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
                myAllLookupStrings, myCaseSensitive);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder strikeout() {
        return withStrikeoutness(true);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withStrikeoutness(boolean strikeout) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.setStrikeout(strikeout);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
                myAllLookupStrings, myCaseSensitive);
    }

    /**
     * @deprecated use {@link #withTailText(String)}
     */
    @Deprecated
    @ApiStatus.ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder setTailText(@Nullable String tailText) {
        return withTailText(tailText);
    }
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withTailText(@Nullable String tailText) {
        return withTailText(tailText, false);
    }

    /**
     * @deprecated use {@link #withTailText(String, boolean)}
     */
    @Deprecated
    @ApiStatus.ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder setTailText(@Nullable String tailText, boolean grayed) {
        return withTailText(tailText, grayed);
    }
    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder withTailText(@Nullable String tailText, boolean grayed) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.setTailText(tailText, grayed);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
                myAllLookupStrings, myCaseSensitive);
    }

    @Contract(pure=true)
    public @NotNull GdLookupElementBuilder appendTailText(@NotNull String tailText, boolean grayed) {
        final LookupElementPresentation presentation = copyPresentation();
        presentation.appendTailText(tailText, grayed);
        return cloneWithUserData(myLookupString, myObject, myInsertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
                myAllLookupStrings, myCaseSensitive);
    }

    @Contract(pure=true)
    public LookupElement withAutoCompletionPolicy(AutoCompletionPolicy policy) {
        return policy.applyPolicy(this);
    }

    @NotNull
    @Override
    public String getLookupString() {
        return myLookupString;
    }

    @Nullable
    public InsertHandler<LookupElement> getInsertHandler() {
        return myInsertHandler;
    }

    @Nullable
    public LookupElementRenderer<LookupElement> getRenderer() {
        return myRenderer;
    }

    @Override
    public @Nullable LookupElementRenderer<? extends LookupElement> getExpensiveRenderer() {
        return myExpensiveRenderer;
    }

    @NotNull
    @Override
    public Object getObject() {
        return myObject;
    }

    @Nullable
    @Override
    public PsiElement getPsiElement() {
        if (myPsiElement != null) return myPsiElement.getElement();
        return super.getPsiElement();
    }

    @Override
    public void handleInsert(@NotNull InsertionContext context) {
        if (myInsertHandler != null) {
            myInsertHandler.handleInsert(context, this);
        }
    }

    @Override
    public void renderElement(LookupElementPresentation presentation) {
        if (myRenderer != null) {
            myRenderer.renderElement(this, presentation);
        }
        else if (myHardcodedPresentation != null) {
            presentation.copyFrom(myHardcodedPresentation);
        } else {
            presentation.setItemText(myLookupString);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        GdLookupElementBuilder that = (GdLookupElementBuilder)o;

        final InsertHandler<LookupElement> insertHandler = that.myInsertHandler;
        if (myInsertHandler != null && insertHandler != null ? !myInsertHandler.getClass().equals(insertHandler.getClass())
                : myInsertHandler != insertHandler) return false;
        if (!myLookupString.equals(that.myLookupString)) return false;
        if (!myObject.equals(that.myObject)) return false;

        final LookupElementRenderer<LookupElement> renderer = that.myRenderer;
        if (myRenderer != null && renderer != null ? !myRenderer.getClass().equals(renderer.getClass()) : myRenderer != renderer) return false;

        return true;
    }

    @Override
    public String toString() {
        return "GdLookupElementBuilder: string=" + getLookupString() + "; handler=" + myInsertHandler;
    }

    @Override
    public int hashCode() {
        int result = 0;
        result = 31 * result + (myInsertHandler != null ? myInsertHandler.getClass().hashCode() : 0);
        result = 31 * result + (myLookupString.hashCode());
        result = 31 * result + (myObject.hashCode());
        result = 31 * result + (myRenderer != null ? myRenderer.getClass().hashCode() : 0);
        return result;
    }

}

File: gdscript/src/main/kotlin/gdscript/completion/utils/GdLookupElementBuilder.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package gdscript.competion.utils

import com.intellij.codeInsight.completion.InsertHandler
import com.intellij.codeInsight.completion.InsertionContext
import com.intellij.codeInsight.lookup.AutoCompletionPolicy
import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.lookup.LookupElementPresentation
import com.intellij.codeInsight.lookup.LookupElementRenderer
import com.intellij.openapi.util.text.StringUtil
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.SmartPointerManager
import com.intellij.psi.SmartPsiElementPointer
import com.intellij.psi.util.PsiUtilCore
import org.jetbrains.annotations.ApiStatus.ScheduledForRemoval
import org.jetbrains.annotations.Contract
import java.awt.Color
import java.util.Collections
import javax.swing.Icon

/**
 * Copied directly from com.intellij.codeInsight.lookup.LookupElementBuilder
 * The only change is a new create() method with different lookup string & object -> a hack to add extra String for partial text match completions
 *
 * @author peter
 *
 * @see com.intellij.codeInsight.completion.PrioritizedLookupElement
 */
class GdLookupElementBuilder private constructor(
    private val myLookupString: String,
    private val myObject: Any,
    val insertHandler: InsertHandler<LookupElement>? = null,
    val renderer: LookupElementRenderer<LookupElement>? = null,
    private val myExpensiveRenderer: LookupElementRenderer<LookupElement>? = null,
    private val myHardcodedPresentation: LookupElementPresentation? = null,
    private val myPsiElement: SmartPsiElementPointer<*>? = null,
    allLookupStrings: Set<String> = setOf(myLookupString),
    caseSensitive: Boolean = true
) : LookupElement() {
    private val myCaseSensitive: Boolean
    private val myAllLookupStrings: Set<String>

    init {
        myAllLookupStrings = Collections.unmodifiableSet(allLookupStrings)
        myCaseSensitive = caseSensitive
    }

    private fun cloneWithUserData(
        lookupString: String, `object`: Any,
        insertHandler: InsertHandler<LookupElement>?,
        renderer: LookupElementRenderer<LookupElement>?,
        expensiveRenderer: LookupElementRenderer<LookupElement>?,
        hardcodedPresentation: LookupElementPresentation?,
        psiElement: SmartPsiElementPointer<*>?,
        allLookupStrings: Set<String>,
        caseSensitive: Boolean
    ): GdLookupElementBuilder {
        val result = GdLookupElementBuilder(
            lookupString, `object`, insertHandler, renderer, expensiveRenderer,
            hardcodedPresentation, psiElement, allLookupStrings, caseSensitive
        )
        copyUserDataTo(result)
        return result
    }

    @ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure = true)
    @Deprecated("use {@link #withInsertHandler(InsertHandler)}")
    fun setInsertHandler(insertHandler: InsertHandler<LookupElement>?): GdLookupElementBuilder {
        return withInsertHandler(insertHandler)
    }

    @Contract(pure = true)
    fun withInsertHandler(insertHandler: InsertHandler<LookupElement>?): GdLookupElementBuilder {
        return cloneWithUserData(
            myLookupString, myObject, insertHandler,
            renderer, myExpensiveRenderer, myHardcodedPresentation,
            myPsiElement, myAllLookupStrings, myCaseSensitive
        )
    }

    @Contract(pure = true)
    fun withRenderer(renderer: LookupElementRenderer<LookupElement>?): GdLookupElementBuilder {
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler, renderer, myExpensiveRenderer, myHardcodedPresentation,
            myPsiElement, myAllLookupStrings, myCaseSensitive
        )
    }

    @Contract(pure = true)
    fun withExpensiveRenderer(expensiveRenderer: LookupElementRenderer<LookupElement>?): GdLookupElementBuilder {
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler,
            renderer, expensiveRenderer, myHardcodedPresentation,
            myPsiElement, myAllLookupStrings, myCaseSensitive
        )
    }

    override fun getAllLookupStrings(): Set<String> {
        return myAllLookupStrings
    }

    @ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure = true)
    @Deprecated("use {@link #withIcon(Icon)}")
    fun setIcon(icon: Icon?): GdLookupElementBuilder {
        return withIcon(icon)
    }

    @Contract(pure = true)
    fun withIcon(icon: Icon?): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.icon = icon
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
            myAllLookupStrings, myCaseSensitive
        )
    }

    private fun copyPresentation(): LookupElementPresentation {
        val presentation = LookupElementPresentation()
        if (myHardcodedPresentation != null) {
            presentation.copyFrom(myHardcodedPresentation)
        } else {
            presentation.itemText = myLookupString
        }
        return presentation
    }

    @Contract(pure = true)
    fun withLookupString(another: String): GdLookupElementBuilder {
        val set: MutableSet<String> = HashSet(myAllLookupStrings)
        set.add(another)
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler,
            renderer, myExpensiveRenderer, myHardcodedPresentation,
            myPsiElement, Collections.unmodifiableSet(set), myCaseSensitive
        )
    }

    @Contract(pure = true)
    fun withLookupStrings(another: Collection<String>): GdLookupElementBuilder {
        val set: MutableSet<String> = HashSet(myAllLookupStrings.size + another.size)
        set.addAll(myAllLookupStrings)
        set.addAll(another)
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler,
            renderer, myExpensiveRenderer, myHardcodedPresentation,
            myPsiElement, Collections.unmodifiableSet(set), myCaseSensitive
        )
    }

    override fun isCaseSensitive(): Boolean {
        return myCaseSensitive
    }

    /**
     * @param caseSensitive if this lookup item should be completed in the same letter case as prefix
     * @return modified builder
     * @see com.intellij.codeInsight.completion.CompletionResultSet.caseInsensitive
     */
    @Contract(pure = true)
    fun withCaseSensitivity(caseSensitive: Boolean): GdLookupElementBuilder {
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler,
            renderer, myExpensiveRenderer, myHardcodedPresentation,
            myPsiElement, myAllLookupStrings, caseSensitive
        )
    }

    /**
     * Allows to pass custom PSI that will be returned from [.getPsiElement].
     */
    @Contract(pure = true)
    fun withPsiElement(psi: PsiElement?): GdLookupElementBuilder {
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler,
            renderer, myExpensiveRenderer, myHardcodedPresentation,
            if (psi == null) null else SmartPointerManager.createPointer(psi),
            myAllLookupStrings, myCaseSensitive
        )
    }

    @Contract(pure = true)
    fun withItemTextForeground(itemTextForeground: Color): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.itemTextForeground = itemTextForeground
        return cloneWithUserData(
            myLookupString, myObject, insertHandler, null, myExpensiveRenderer, presentation,
            myPsiElement, myAllLookupStrings, myCaseSensitive
        )
    }

    @Contract(pure = true)
    fun withItemTextUnderlined(underlined: Boolean): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.isItemTextUnderlined = underlined
        return cloneWithUserData(
            myLookupString, myObject, insertHandler, null, myExpensiveRenderer, presentation,
            myPsiElement, myAllLookupStrings, myCaseSensitive
        )
    }

    @Contract(pure = true)
    fun withItemTextItalic(italic: Boolean): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.isItemTextItalic = italic
        return cloneWithUserData(
            myLookupString, myObject, insertHandler, null, myExpensiveRenderer, presentation,
            myPsiElement, myAllLookupStrings, myCaseSensitive
        )
    }

    @ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure = true)
    @Deprecated("use {@link #withTypeText(String)}")
    fun setTypeText(typeText: String?): GdLookupElementBuilder {
        return withTypeText(typeText)
    }

    @Contract(pure = true)
    fun withTypeText(typeText: String?): GdLookupElementBuilder {
        return withTypeText(typeText, false)
    }

    @Contract(pure = true)
    fun withTypeText(typeText: String?, grayed: Boolean): GdLookupElementBuilder {
        return withTypeText(typeText, null, grayed)
    }

    @Contract(pure = true)
    fun withTypeText(typeText: String?, typeIcon: Icon?, grayed: Boolean): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.setTypeText(typeText, typeIcon)
        presentation.isTypeGrayed = grayed
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
            myAllLookupStrings, myCaseSensitive
        )
    }

    fun withTypeIconRightAligned(typeIconRightAligned: Boolean): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.isTypeIconRightAligned = typeIconRightAligned
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
            myAllLookupStrings, myCaseSensitive
        )
    }

    @ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure = true)
    @Deprecated("use {@link #withPresentableText(String)}")
    fun setPresentableText(presentableText: String): GdLookupElementBuilder {
        return withPresentableText(presentableText)
    }

    @Contract(pure = true)
    fun withPresentableText(presentableText: String): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.itemText = presentableText
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
            myAllLookupStrings, myCaseSensitive
        )
    }

    @Contract(pure = true)
    fun bold(): GdLookupElementBuilder {
        return withBoldness(true)
    }

    @Contract(pure = true)
    fun withBoldness(bold: Boolean): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.isItemTextBold = bold
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
            myAllLookupStrings, myCaseSensitive
        )
    }

    @Contract(pure = true)
    fun strikeout(): GdLookupElementBuilder {
        return withStrikeoutness(true)
    }

    @Contract(pure = true)
    fun withStrikeoutness(strikeout: Boolean): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.isStrikeout = strikeout
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
            myAllLookupStrings, myCaseSensitive
        )
    }

    @ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure = true)
    @Deprecated("use {@link #withTailText(String)}")
    fun setTailText(tailText: String?): GdLookupElementBuilder {
        return withTailText(tailText)
    }

    @Contract(pure = true)
    fun withTailText(tailText: String?): GdLookupElementBuilder {
        return withTailText(tailText, false)
    }

    @ScheduledForRemoval(inVersion = "2021.3")
    @Contract(pure = true)
    @Deprecated("use {@link #withTailText(String, boolean)}")
    fun setTailText(tailText: String?, grayed: Boolean): GdLookupElementBuilder {
        return withTailText(tailText, grayed)
    }

    @Contract(pure = true)
    fun withTailText(tailText: String?, grayed: Boolean): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.setTailText(tailText, grayed)
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
            myAllLookupStrings, myCaseSensitive
        )
    }

    @Contract(pure = true)
    fun appendTailText(tailText: String, grayed: Boolean): GdLookupElementBuilder {
        val presentation = copyPresentation()
        presentation.appendTailText(tailText, grayed)
        return cloneWithUserData(
            myLookupString, myObject,
            insertHandler, null, myExpensiveRenderer, presentation, myPsiElement,
            myAllLookupStrings, myCaseSensitive
        )
    }

    @Contract(pure = true)
    fun withAutoCompletionPolicy(policy: AutoCompletionPolicy): LookupElement {
        return policy.applyPolicy(this)
    }

    override fun getLookupString(): String {
        return myLookupString
    }

    override fun getExpensiveRenderer(): LookupElementRenderer<out LookupElement>? {
        return myExpensiveRenderer
    }

    override fun getObject(): Any {
        return myObject
    }

    override fun getPsiElement(): PsiElement? {
        return if (myPsiElement != null) myPsiElement.element else super.getPsiElement()
    }

    override fun handleInsert(context: InsertionContext) {
        if (insertHandler != null) {
            insertHandler.handleInsert(context, this)
        }
    }

    override fun renderElement(presentation: LookupElementPresentation) {
        if (renderer != null) {
            renderer.renderElement(this, presentation)
        } else if (myHardcodedPresentation != null) {
            presentation.copyFrom(myHardcodedPresentation)
        } else {
            presentation.itemText = myLookupString
        }
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as GdLookupElementBuilder
        val insertHandler = that.insertHandler
        if (if (this.insertHandler != null && insertHandler != null) insertHandler.javaClass != insertHandler.javaClass else this.insertHandler !== insertHandler) return false
        if (myLookupString != that.myLookupString) return false
        if (myObject != that.myObject) return false
        val renderer = that.renderer
        return if (if (this.renderer != null && renderer != null) renderer.javaClass != renderer.javaClass else this.renderer !== renderer) false else true
    }

    override fun toString(): String {
        return "GdLookupElementBuilder: string=" + lookupString + "; handler=" + insertHandler
    }

    override fun hashCode(): Int {
        var result = 0
        result = 31 * result + if (insertHandler != null) insertHandler.javaClass.hashCode() else 0
        result = 31 * result + myLookupString.hashCode()
        result = 31 * result + myObject.hashCode()
        result = 31 * result + if (renderer != null) renderer.javaClass.hashCode() else 0
        return result
    }

    companion object {
        fun create(lookupString: String): GdLookupElementBuilder {
            return GdLookupElementBuilder(lookupString, lookupString)
        }

        fun create(lookupString: String, lookupObject: String): GdLookupElementBuilder {
            return GdLookupElementBuilder(lookupString, lookupObject)
        }

        fun create(`object`: Any): GdLookupElementBuilder {
            return GdLookupElementBuilder(`object`.toString(), `object`)
        }

        fun createWithSmartPointer(lookupString: String, element: PsiElement): GdLookupElementBuilder {
            PsiUtilCore.ensureValid(element)
            return GdLookupElementBuilder(
                lookupString,
                SmartPointerManager.getInstance(element.project).createSmartPsiElementPointer(element)
            )
        }

        fun create(element: PsiNamedElement): GdLookupElementBuilder {
            PsiUtilCore.ensureValid(element)
            return GdLookupElementBuilder(StringUtil.notNullize(element.name), element)
        }

        fun createWithIcon(element: PsiNamedElement): GdLookupElementBuilder {
            PsiUtilCore.ensureValid(element)
            return create(element).withIcon(element.getIcon(0))
        }

        fun create(lookupObject: Any, lookupString: String): GdLookupElementBuilder {
            if (lookupObject is PsiElement) {
                PsiUtilCore.ensureValid(lookupObject)
            }
            return GdLookupElementBuilder(lookupString, lookupObject)
        }
    }
}